### PR TITLE
Linux support and CI

### PR DESCRIPTION
**Pull request by gorazdko**

### CODE REVIEW

**Comment by gorazdko** [comment]
@wolfmcnally can you review and merge

Needed, see https://github.com/BlockchainCommons/bc-sskr/issues/3#issuecomment-789092258

**Comment by AndreasGassmann** [comment]
If this is the expected output, then I can confirm that this fix works for me:

```
vscode ➜ /workspaces/bc-lifehash (linux-support ✗) $ make
cd src && make all
make[1]: Entering directory '/workspaces/bc-lifehash/src'
ar rcs -o liblifehash.a bit-aggregator.o bit-enumerator.o cell-grid.o color.o color-func.o color-grid.o format-utils.o gradients.o hsb-color.o lifehash.o memzero.o patterns.o point.o sha256.o
make[1]: Leaving directory '/workspaces/bc-lifehash/src'
cd test && make all
make[1]: Entering directory '/workspaces/bc-lifehash/test'
cc -g -O0   -c -o testc.o testc.c
cc   testc.o ../src/liblifehash.a  -L../src -llifehash -lstdc++ -lm -o testc
g++    -c -o test.o test.cpp
cc   test.o ../src/liblifehash.a  -L../src -llifehash -lstdc++ -lm -o test
make[1]: Leaving directory '/workspaces/bc-lifehash/test'
```

**Comment by gorazdko** [comment]
> If this is the expected output, then I can confirm that this fix works for me

yes it's correct. If you also run `make check` the tests will be run.

File: build.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!zsh
```
Comment (wolfmcnally) [inline build.sh:1]: My build scripts for MacOS now routinely rely on features of ZSH, which is now the Mac-standard shell. I don't guarantee that things will build on the Mac if you try to switch back to using Bash.

File: tasks.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!zsh
```
Comment (wolfmcnally) [inline tasks.sh:1]: See previous comment about using ZSH vs. Bash.

**Review by wolfmcnally — CHANGES_REQUESTED** [review]
The build.sh script should be run under ZSH, not Bash.

**Comment by ChristopherA** [comment]
Shouldn't both zsh and bash do the right thing if the script has the header info correct? 

`#!/bin/zsh`

`#!/bin/bash`

**Comment by wolfmcnally** [comment]
> Shouldn't both zsh and bash do the right thing if the script has the header info correct?
> 
> `#!/bin/zsh`
> 
> `#!/bin/bash`

Bash and Zsh are very different in some important ways. They are not simply interchangeable. I have switched to Zsh for writing build scripts, especially because dealing with file paths under Zsh is actually a lot more straightforward, and I'm not going to guarantee or approve scripts I've written and tested under Zsh being slammed over to Bash.

**Comment by gorazdko** [comment]
Closing as I will not be supporting zsh with Linux

**Comment by wolfmcnally** [comment]
> Closing as I will not be supporting zsh with Linux

Why not? It's only for the build script. And all my build scripts are going to be migrated to Zsh eventually because they're far easier to maintain.

**Comment by ChristopherA** [comment]
I believe (but don't know) that Debian and most distributions come with zsh now. Mac one has a very old bash and refuses to update it. If you were to tell me that Debian didn't have curren tzsh, I'd say we'd have to support the least common denominator.

**Comment by gorazdko** [comment]
The majority of devices/platforms do not come preinstalled with zsh, they do come with bash.

I've added new commits to support zsh. I dont see the benfits of zsh outweighing the costs though (changing all the repos, setting windows cross compiler, new bugs, is the change easy to adopt when cross compiling to xyz platform)

But I'll let you guys decide.

> And all my build scripts are going to be migrated to Zsh eventually because they're far easier to maintain.

If you're going to do that make sure you're adding the extra 2 lines (these: https://github.com/BlockchainCommons/bc-lifehash/pull/2/commits/17fb855541565120b774a5434a4dd47084c13b9b) in every repo you are changing to make the linux build pass.

**Review by wolfmcnally — APPROVED** [review]
LGTM

**Comment by wolfmcnally** [comment]
@gorazdko Just so you know I have my reasons: not every script cares about this, but I've run across it enough that for my own coding (and promotion of better tech) I'm choosing Zsh over Bash, primarily for these reasons excerpted below:

- [What are the practical differences between bash and zsh?](https://apple.stackexchange.com/questions/361870/what-are-the-practical-differences-between-bash-and-zsh)
    - In bash, $foo takes the value of foo, splits it at whitespace characters, and for each whitespace-separated part, if it contains wildcard characters and matches an existing file, replaces the pattern by the list of matches. To just get the value of foo, you need "$foo". The same applies to command substitution $(foo). In zsh, $foo is the value of foo and $(foo) is the output of foo minus its final newlines, with two exceptions. If a word becomes empty due to expanding empty unquoted variables, it's removed (e.g. a=; b=; printf "%s\n" one "$a$b" three $a$b five prints one, an empty line, three, five). The result of an unquoted command substitution is split at whitespace but the pieces don't undergo wildcard matching.
- [Why does my shell script choke on whitespace or other special characters?
](https://unix.stackexchange.com/questions/131766/why-does-my-shell-script-choke-on-whitespace-or-other-special-characters)
    - "To process a list of any kind (such as file names), put it in an array. This requires mksh, ksh93, yash or bash (or **zsh**, which doesn't have all these quoting issues); a plain POSIX shell (such as ash or dash) doesn't have array variables."

**Pull request merged**